pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/SchemaChangeHandler.java

```diff
@@ -0,0 +1,341 @@
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.linkedin.pinot.common.data.DimensionFieldSpec;
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.common.data.Schema;
+import com.linkedin.pinot.common.data.FieldSpec.FieldType;
+import com.linkedin.pinot.common.metadata.segment.IndexLoadingConfigMetadata;
+import com.linkedin.pinot.common.segment.ReadMode;
+import com.linkedin.pinot.core.segment.creator.ColumnIndexCreationInfo;
+import com.linkedin.pinot.core.segment.creator.ForwardIndexType;
+import com.linkedin.pinot.core.segment.creator.InvertedIndexType;
+import com.linkedin.pinot.core.segment.creator.impl.SegmentColumnarIndexCreator;
+import com.linkedin.pinot.core.segment.creator.impl.SegmentDictionaryCreator;
+import com.linkedin.pinot.core.segment.creator.impl.V1Constants;
+import com.linkedin.pinot.core.segment.creator.impl.fwd.MultiValueUnsortedForwardIndexCreator;
+import com.linkedin.pinot.core.segment.creator.impl.fwd.SingleValueSortedForwardIndexCreator;
+import com.linkedin.pinot.core.segment.index.ColumnMetadata;
+import com.linkedin.pinot.core.segment.index.SegmentMetadataImpl;
+import com.linkedin.pinot.core.segment.memory.PinotDataBuffer;
+import com.linkedin.pinot.core.segment.store.ColumnIndexType;
+import com.linkedin.pinot.core.segment.store.SegmentDirectory;
+import com.linkedin.pinot.core.segment.store.SegmentDirectory.Writer;
+
+public class SchemaChangeHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SchemaChangeHandler.class);
+
+  private File indexDir;
+  private SegmentMetadataImpl metadata;
+  private Optional<IndexLoadingConfigMetadata> indexConfig;
+  private Schema schema;
+  private SegmentDirectory segmentDirectory;
+
+  enum Action {
+    ADD_DIMENSION,
+    ADD_METRIC, // present in schema but not in segment
+    UPDATE_DIMENSION_DEFAULT_VALUE,
+    UPDATE_METRIC_DEFAULT_VALUE // present in schema & segment but default values don't match
+  }
+
+  SchemaChangeHandler(File indexDir, SegmentMetadataImpl metadata,
+      IndexLoadingConfigMetadata indexConfig, Schema schema) {
+    Preconditions.checkNotNull(indexDir);
+    Preconditions.checkNotNull(metadata);
+    Preconditions.checkState(indexDir.exists(), "Segment directory: {} does not exist", indexDir);
+    Preconditions.checkState(indexDir.isDirectory(), "Segment path: {} is not a directory",
+        indexDir);
+
+    this.indexDir = indexDir;
+    this.metadata = metadata;
+    this.indexConfig = Optional.fromNullable(indexConfig);
+    this.schema = schema;
+    // always use mmap. That's safest and performs well without impact from
+    // -Xmx params
+    // This is not final load of the segment
+    segmentDirectory = SegmentDirectory.createFromLocalFS(indexDir, metadata, ReadMode.mmap);
```

Comment:
Check if schema is not null before loading? (Loading should be expensive I guess.)